### PR TITLE
Global exception handler and validation

### PR DESCRIPTION
**Pull request by denysokhrymovych**

### CODE REVIEW

File: src/main/java/com/example/bookshop/service/BookServiceImpl.java [hunk 1]
```diff
@@ -34,6 +34,29 @@ public BookDto getBookById(Long id) {
...
+
+        book.setTitle(requestDto.getTitle());
+        book.setAuthor(requestDto.getAuthor());
+        book.setIsbn(requestDto.getIsbn());
+        book.setPrice(requestDto.getPrice());
+        book.setDescription(requestDto.getDescription());
+        book.setCoverImage(requestDto.getCoverImage());
+
+        return bookMapper.toDto(bookRepository.save(book));
+    }
```
Comment (HovorukhaBohdan) [inline src/main/java/com/example/bookshop/service/BookServiceImpl.java]: I think you can change `id` in `book` to `id` from `requestDto`

**Review by HovorukhaBohdan — APPROVED** [review]
Good job, just 1 minor point

File: src/main/java/com/example/bookshop/service/BookServiceImpl.java (hunk 1 above)
Comment (bhdnchui) [reply src/main/java/com/example/bookshop/service/BookServiceImpl.java]: @HovorukhaBohdan Please, pay attention that CreateBookRequestDto doesn't have field `id`.

File: src/main/java/com/example/bookshop/model/Book.java [hunk 2]
```diff
@@ -9,11 +9,16 @@
...
 
-@Entity
-@Table(name = "books")
 @Getter
 @Setter
+@Entity
+@SQLDelete(sql = "UPDATE books SET is_deleted = true WHERE id=?")
+@Where(clause = "is_deleted=false")
+@Table(name = "books")
+
```
Comment (bhdnchui) [inline src/main/java/com/example/bookshop/model/Book.java]:
```suggestion
```

File: src/main/java/com/example/bookshop/repository/BookRepository.java [hunk 3]
```diff
@@ -1,13 +1,8 @@
 package com.example.bookshop.repository;
 
 import com.example.bookshop.model.Book;
-import java.util.List;
-import java.util.Optional;
+import org.springframework.data.jpa.repository.JpaRepository;
 
-public interface BookRepository {
-    List<Book> getAll();
```
Comment (bhdnchui) [inline src/main/java/com/example/bookshop/repository/BookRepository.java]: Add `@Repository` annotation here.

File: src/main/java/com/example/bookshop/service/BookServiceImpl.java [hunk 4]
```diff
@@ -34,6 +34,29 @@ public BookDto getBookById(Long id) {
...
+    public BookDto updateBook(Long id, CreateBookRequestDto requestDto) {
+        Book book = bookRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("Book not found with id: " + id));
+
+        book.setTitle(requestDto.getTitle());
+        book.setAuthor(requestDto.getAuthor());
+        book.setIsbn(requestDto.getIsbn());
+        book.setPrice(requestDto.getPrice());
+        book.setDescription(requestDto.getDescription());
+        book.setCoverImage(requestDto.getCoverImage());
```
Comment (bhdnchui) [inline src/main/java/com/example/bookshop/service/BookServiceImpl.java]: You can minimize lines of your code with replacing this line with the next ones:
```suggestion
        book = bookMapper.toEntity(bookDto);
        book.setId(id);
```
It is not critical, but it is good to do.

File: src/main/java/com/example/bookshop/validation/Isbn.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import org.hibernate.validator.internal.constraintvalidators.hv.ISBNValidator;
+
+@Constraint(validatedBy = ISBNValidator.class)
+@Target({ElementType.PARAMETER, ElementType.FIELD})
+@Retention(RetentionPolicy.RUNTIME)
+public @interface Isbn {
```
Comment (bhdnchui) [inline src/main/java/com/example/bookshop/validation/Isbn.java]: Let's use default org.hibernate.validator.constraints.ISBN annotation for validation.

**Review by bhdnchui — CHANGES_REQUESTED** [review]
Good job!
Just a few comments to process.

File: src/main/java/com/example/bookshop/dto/CreateBookRequestDto.java [hunk 6]
```diff
@@ -1,14 +1,30 @@
...
+    @NotNull(message = "Author cannot be null")
+    @Size(max = 255, message = "Author cannot be greater than 255 characters")
     private String author;
+
+    @NotNull(message = "ISBN cannot be null")
+    @ISBN(message = "Invalid format of ISBN")
     private String isbn;
+
+    @NotNull(message = "Price cannot be null")
+    @Min(0)
```
Comment (bhdnchui) [inline src/main/java/com/example/bookshop/dto/CreateBookRequestDto.java:25]: It's better to use `@Positive` here.

**Review by bhdnchui — APPROVED** [review]
Good job!

**Pull request merged**